### PR TITLE
NIFI-5572

### PR DESCRIPTION
**Pull request by gideonkorir**
Add Decimal support

- RecordFieldType.DECIMAL
- DecimalDataType
- Fix conversions for Avro + ORC
- Fix broken tests

### CODE REVIEW

**Comment by MikeThomsen** [comment]
@gideonkorir can you rebase and force push to your branch?

**Comment by gideonkorir** [comment]
@MikeThomsen I've just done that again after fixing the style issues

**Comment by joewitt** [comment]
@gideonkorir there is a conflict and ideally the commits would be squashed to allow for reviews to result in easier to follow changes.

**Comment by gideonkorir** [comment]
@joewitt I've fixed the conflicts, I assumed it's a squash merge according to [contributor guide](https://cwiki.apache.org/confluence/display/NIFI/Contributor+Guide#ContributorGuide-RebasingandSquashingwhileunderreview).

**Comment by gideonkorir** [comment]
Realized I was naive in my initial PR, I've gone through all the references to `RecordFieldType` and made the following changes:

- [x]  Update PutCassandra processor to handle decimal and added tests for that
- [x] Update PutElasticSearch processor to handle decimal and added tests for that. Caveat, I'm using BigDecimal.toPlainString() so that we don't have scientific notation in the json output; not sure if OK
- [x] Added decimal tests for HBase
- [x] Added decimal tests for Kudu
- [x] Update PutMongoRecord processor to handle decimal, added codec package that adds the necessary codecs for that. 
- [x] Update StandardSchemaValidator, support decimal in `validate`, added tests for that
- [x] ResultSetRecordSet has decimal support, factoring in precision and scale of the returned decimals
- [x] JsonPathRowRecordReader has decimal support, added tests for that
- [x] JsonTreeRowRecordReader has decimal support, added tests for that

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/RecordFieldType.java [hunk 1]
```diff
@@ -265,6 +270,19 @@ public DataType getDataType(final String format) {
         return new DataType(this, format);
     }
 
+    /**
+     * Returns a Data Type that represents a "Decimal" with given precision
+     * @param precision number of digits to the right of point (.)
```
Comment (matthew-formosa-gig) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/RecordFieldType.java]: The scale is actually the number of digits to the right of the point. The precision is the number of digits in the number.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java [hunk 2]
```diff
@@ -172,6 +180,8 @@ public static Object convertType(final Object value, final DataType dataType, fi
 
                 return convertType(value, chosenDataType, fieldName, charset);
             }
+            case DECIMAL:
+                return toDecimal(value, fieldName).setScale(((DecimalDataType)dataType).getScale(), RoundingMode.HALF_UP);
```
Comment (matthew-formosa-gig) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java]: Why not pass the scale (and ideally precision) to the `toDecimal` method and set the scale and rounding mode (and check precision) within that method?

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java [hunk 3]
```diff
@@ -1378,6 +1393,61 @@ public static boolean isCharacterTypeCompatible(final Object value) {
...
+        }
+
+        if (value instanceof String) {
+            return new BigDecimal((String) value);
+        }
+
+        if(value instanceof Number){
+            return new BigDecimal(((Number)value).doubleValue());
+        }
+
```
Comment (matthew-formosa-gig) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java]: I would also check for precision here and throw an `IllegalTypeConversionException` if the required precision is less than the precision of the converted `BigDecimal`.

File: nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/resources/json/output/dataTypes.json [hunk 4]
```diff
@@ -11,6 +11,7 @@
   "time" : "17:00:00",
   "timestamp" : "2017-01-01 17:00:00",
   "char" : "c",
+  "decimal" : "45.76",
```
Comment (matthew-formosa-gig) [inline nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/resources/json/output/dataTypes.json]: I would rather have this as a number to be honest. There is a specific case in the `WriteJsonResult` class.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/RecordFieldType.java (hunk 1 above)
Comment (gideonkorir) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/RecordFieldType.java]: You right don't know what I was thinking when I wrote the comment, will fix that. thanks

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java (hunk 3 above)
Comment (gideonkorir) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java]: Good feedback, will add that

**Pull request closed without merging**